test/web-platform-tests/to-run.yaml

```diff
@@ -135,7 +135,10 @@ Element-insertAdjacentElement.html: [fail, Not implemented]
 Element-insertAdjacentText.html: [fail, Not implemented]
 Element-matches.html: [fail, Whitespace in pseudo-class parameter]
 Element-webkitMatchesSelector.html: [fail, Whitespace in pseudo-class parameter]
-MutationObserver-**: [fail, Not implemented]
+MutationObserver-characterData.html: [fail, Range is not implemented]
+MutationObserver-childList.html: [fail, Range is not implemented]
+MutationObserver-document.html: [fail, Compound microtask scheduling issue]
+MutationObserver-inner-outer.html: [fail, innerHTML and outerHTML are not spec compliant]
```

Comment:
In what way is this still true?

Reply:
The `innerHTML` and `outerHTML` are directly manipulating the context object, using [`appendChild`](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/browser/htmltodom.js#L58-L65), [`removeChild`](https://github.com/jsdom/jsdom/blob/master/lib/jsdom/living/helpers/html.js#L6-L8). If a mutation observer is attached to the tree, those mutations will enqueue new records.

As far as I understand [Element.innerHTML](https://w3c.github.io/DOM-Parsing/#dfn-innerhtml), the parser should produce a document fragment and execute the replace all algorithm on the context object. Same goes for `Element.outerHTML`.

It will require a large amount of refactor to the `htmltodom.js`, and I think it would deserve another PR. Here is an ugly draft of what I have in mind: https://github.com/pmdartus/jsdom/compare/impl-mutation-observer...pmdartus:inner-outer-html-parsing. The code `htmltodom.js` is duplicated, but at least `Element.innerHTML` and `Element.outerHTML` are conform to the spec. Those changes make `MutationObserver-inner-outer.html` pass.

I would be happy to wait for the parsing aglo to be fixed before merging this PR.

Reply:
Ah OK, makes sense! I'm happy to land fixes for innerHTML/outerHTML as a follow-up; it seems like a clean separation. Although I hope you have time to contribute them as well even after we land MutationObserver, as conformant innerHTML/outerHTML would be another great improvement :).

Reply:
Yes, will do!